src/lanco.8
.\" Copyright (c) 2013 Vincent Bernat <vincent.bernat@dailymotion.com>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: February 05 2013 $
.Dt LANCO 8
.Os
.Sh NAME
.Nm lanco
.Nd task launcher
.Sh SYNOPSIS
.Nm
.Op Fl dv
.Op Fl D Ar debug
.Ar namespace
.Ar command
.Op ...
.Sh DESCRIPTION
.Nm
is a task launcher which does not want to be an init program. It can
run any task, stop them and check if they are still running. It
leverages cgroups in recent Linux kernels by enclosing each task in a
dedicated cgroup.
.Pp
.Nm
does not have any daemon.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Be more verbose. This option can be repeated twice to enable debug
mode. Debug messages can then be filtered with
.Fl D
flag.
.It Fl D Ar debug
This option allows the user to filter out debugging information by
specifying allowed tokens. This option can be repeated several times
to allow several tokens. This option must be combined with the
.Fl d
flag to have some effect. Only debugging logs can be filtered.
.El
.Nm
expects to be called with a namespace. Namespaces allow a user to use
.Nm
for several different usages. After a namespace, a command is
expected. Each command has its own set of options described below.
.Sh COMMANDS
.Nm
accepts the following commands:

.Cd init
.Op Fl u Ar user
.Op Fl g Ar group
.Bd -ragged -offset XX
Initialize the namespace for use by the given user and group. This
commands is mandatory to be able to use the other commands with the
given namespace. It also allows one to run subsequent
.Nm
commands without special privileges.
.Ed

.Cd run
.Op Fl f
.Op Fl L
.Op Fl l Ar logfile
.Op Fl c Ar command
.Op Fl m Ar limit
.Ar taskname
.Ar command
.Bd -ragged -offset XX
Run the provided command in the namespace as task
.Ar taskname .
The command will fail if the task is already running. If the
.Fl f
flag is provided, the command will be run in foreground. Otherwise,
the standard input will be closed, the standard output and the
standard error are redirected to a file
.Ar logfile
or to
.Pa /var/log/lanco-XXXXX/YYYYY.log
where the first pattern is the namespace and the second is the task
name. If the log file already exists, it will be rotated. This
behaviour can also be obtained when the command is run in foreground
by specifying
.Fl L .
.Pp
With the
.Fl c
flag,
.Nm
will run the provided command when the task exits. This allows one to
be notified of the task completion. The command is run by
.Pa /bin/sh .
.Pp
On systems where memory cgroup is available, it is possible to limit
the memory usage of a task by using the
.Fl m
flag. The limit should be set in bytes. On most kernels, memory cgroup
can be enabled by passing
.Cm cgroup_enable=memory
to the kernel.
.Ed

.Cd stop
.Ar taskname
.Bd -ragged -offset XX
Stop the provided task. The command will fail if the task is not
running. A process can evade this command by creating a sub-cgroup in
the named hierarchy. This should not happen since the hierarchy is
reserved for
.Nm
use.
.Ed

.Cd release
.Op Fl n
.Op Ar taskname
.Bd -ragged -offset XX
Release the provided task. If no task is provided, release the whole
cgroup. For a task to be released, it should not be running. In this
case, it should have been released automatically. This command is
called internally for this purpose. For the whole namespace to be
released, no task should be present. This command does not clean the
log files. The
.Fl n
flag disables execution of registered command to be run on task
exit. It does not have any effect if no task is provided.
.Ed

.Cd check
.Ar taskname
.Bd -ragged -offset XX
Check if the provided task is already running. The exit status will be
0 if the task is running and 1 otherwise.
.Ed

.Cd ls
.Op Fl l
.Bd -ragged -offset XX
Show all tasks running. By default, the command is truncated to 50
characters. With
.Fl l ,
no truncation occurs.
.Ed

.Cd top
.Bd -ragged -offset XX
Show all tasks running in a top-like output with consumed CPU and
number of tasks. Auto-refresh.
.Ed

.Cd dump
.Bd -ragged -offset XX
Dump all known information about a namespace in JSON format. This
includes the number of tasks, the CPU usage, the number of CPU and for
each task, the list of processes running in the task and the CPU usage
of the task. The CPU usage is the number of nanoseconds per CPU spent
on the task.
.Ed

.Sh ENVIRONMENT
It is expected that
.Nm
is run in a kernel supporting cgroups. If your kernel supports
cgroups, the
.Pa /sys/fs/cgroups
directory exists. Usually, this directory is then populated by
individual controllers. Usually, you get one directory for each of
them. For example,
.Pa /sys/fs/cgroups/memory
is the root of a hierarchy controlled by the memory subsystem.
.Pp
.Nm
can take advantage of cpu accounting subsystem to track CPU usage of
each task. Usually, this subsystem is joined with the cpu
subsystem. This means you should have a
.Pa /sys/fs/cgroups/cpu,cpuacct
directory which is a hierarchy controlled by both the cpu subsystem
and the cpu accounting subsystem. If you have
.Pa /sys/fs/cgroups/cpuacct
directory or a
.Pa /sys/fs/cgroups/cpu,cpuacct
directory, all should work fine. If not,
.Nm
will try to create
.Pa /sys/fs/cgroups/cpu,cpuacct
during initialization of the first namespace. It will also create a
symlink
.Pa /sys/fs/cgroups/cpuacct .
This mimics the expected usage.
.Pp
If the creation of the cpu accounting subsystem fail,
.Nm
won't be able to track CPU usage. This may happen if you have all
subsystems mounted in
.Pa /sys/fs/cgroups .
You should fix your system to not do that. Mounting all subsystems in
the same hierarchy may enable subsystems that need user input to be
functional, like cpuset.

.Sh FILES
.Bl -tag -width "/sys/fs/cgroup/lanco-XXXXX" -compact
.It /sys/fs/cgroup/lanco-XXXXX
Named cgroup for a given namespace.
.It /var/log/lanco-XXXXX/YYYYYYY.log
Log file for a given task in a given namespace. Those files are
automatically rotated.
.It /var/run/lanco-XXXXX/@release-agent
Symbolic link to
.Nm
to act as a release agent for cgroups.
.El

.Sh AUTHORS
.An -nosplit
The
.Nm
program was written by
.An Vincent Bernat Aq vincent.bernat@dailymotion.com .


src/lanco.c
/* -*- mode: c; c-file-style: "openbsd" -*- */
/*
 * Copyright (c) 2013 Vincent Bernat <vincent.bernat@dailymotion.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "lanco.h"

#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <assert.h>

extern const char *__progname;

static void
usage(void)
{
	fprintf(stderr, "Usage: %s [OPTIONS] <namespace> <command> [OPTIONS ...]\n",
		__progname);
	fprintf(stderr, "Version: %s\n", PACKAGE_STRING);
	fprintf(stderr, "\n");
	fprintf(stderr, "-d      Be more verbose.\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "see manual page lanco(8) for more information\n");
}

struct cmd {
	const char *name;
	int(*fn)(const char *, int argc, char * const *argv);
};
static struct cmd lanco_cmds[] = {
	{ "init",    cmd_init },
	{ "run",     cmd_run  },
	{ "start",   cmd_run  },
	{ "stop",    cmd_stop },
	{ "kill",    cmd_stop },
	{ "check",   cmd_check },
	{ "release", cmd_release },
	{ "ls",      cmd_ls },
	{ "top",     cmd_top },
	{ "dump",    cmd_dump },
	{ NULL }
};

/**
 * Expand argv[0] into arguments.
 *
 * lanco can be called with "lanco" but also with embedded arguments,
 * like this:
 *
 * @verbatim
 * lanco@@namespace@@init
 * lanco@@namespace@@run@@task@@sleep 10
 * @endverbatim
 *
 * We expand this here. Other arguments are pushed away.
 */
static void
expand(int *argc, char **argv[])
{
	int count = 0;
	char *ch = (*argv)[0], *pch;
	while (ch && *ch) {
		while (ch && *ch && *(ch+1) != '@') {
			ch++;
			ch = strchr(ch+1, '@');
		}
		if (ch && *ch) {
			count++;
			ch += 2;
		}
	}
	if (count == 0) return;

	/* We need to reallocate a new argv array (NULL terminated, like the
	 * original one) */
	char **nargv = calloc(count + *argc + 1, sizeof(char*));
	assert(nargv);

	/* New arguments */
	count = 0;
	ch = pch = (*argv)[0];
	while (ch && *ch) {
		while (ch && *ch && *(ch+1) != '@') {
			ch++;
			ch = strchr(ch, '@');
		}
		if (ch && *ch) {
			*ch = '\0';
			ch += 2;
			nargv[count++] = pch;
			pch = ch;
		}
	}
	nargv[count] = pch;

	/* Remaining arguments */
	for (int i = 1; i < *argc; i++) {
		nargv[count+i] = (*argv)[i];
	}
	*argc += count;
	*argv = nargv;
}

int
main(int argc, char * const argv[])
{
	int debug = 1;
	int ch;

	/* Do argv[0] expansion */
	expand(&argc, (char***)&argv);

	while ((ch = getopt(argc, argv, "+hvdD:")) != -1) {
		switch (ch) {
		case 'h':
			usage();
			exit(EXIT_SUCCESS);
			break;
		case 'v':
			fprintf(stdout, "%s\n", PACKAGE_VERSION);
			exit(EXIT_SUCCESS);
			break;
		case 'd':
			debug++;
			break;
		case 'D':
			log_accept(optarg);
			break;
		default:
			usage();
			exit(EXIT_FAILURE);
		}
	}

	log_init(debug, __progname);

	/* Namespace and command */
	if (optind > argc - 2) {
		usage();
		exit(EXIT_FAILURE);
	}

	const char *namespace = argv[optind];
	const char *command = argv[optind+1];

	if (!utils_is_valid_name(namespace)) {
		log_warnx("main", "namespace should be alphanumeric ASCII string");
		exit(EXIT_FAILURE);
	}

	log_debug("main", "namespace: %s", namespace);
	log_debug("main", "command: %s", command);
	argc -= optind;
	argv = &argv[optind];
	optind = 1;

	for (struct cmd *cmd = lanco_cmds; cmd->name; cmd++)
		if (!strcmp(cmd->name, command))
			return (cmd->fn(namespace,
				argc - optind,
				&argv[optind]) == 0)?EXIT_SUCCESS:EXIT_FAILURE;

	log_warnx("main", "no command `%s`", command);
	usage();
	exit(EXIT_FAILURE);
}
